src/route.cr

```diff
@@ -112,52 +98,20 @@ module Armature
         end
       end
 
-      def on(*paths : String)
-        paths.each do |path|
-          on(path) { yield }
-        end
-      end
-
-      def on(path : String)
-        return if handled?
-
-        if match?(path)
-          begin
-            old_path = original_request.path
-            original_request.path = original_request.path.sub(/\A\/?#{path}/, "")
-            yield
-          ensure
-            original_request.path = old_path.not_nil!
-          end
-        end
-      end
-
-      def on(capture : Symbol)
+      def on(*segments)
         return if handled?
 
         old_path = original_request.path
-        match = %r(\A/?[^/]+).match original_request.path.sub(%r(\A/), "")
-        if match
-          original_request.path = original_request.path.sub(%r(\A/#{match[0]}), "")
-
-          yield match[0]
-        end
-      ensure
-        if old_path
-          original_request.path = old_path
-        end
-      end
-
-      def on(**capture)
-        return if handled?
-
-        old_path = original_request.path
-        capture.each do |key, value|
-          if (match = %r(\A/?[^/]+).match original_request.path.sub(%r(\A/), "")) && (result = match?(match[0], value))
-            original_request.path = original_request.path.sub(%r(\A/#{match[0]}), "")
-            yield result
+        captures = segments.map do |matcher|
+          if (match = %r(\A/?[^/]+).match original_request.path.sub(%r(\A/), "")) && (result = match?(match[0], matcher))
+            original_request.path = original_request.path.sub(%r(\A/?#{match[0]}), "")
+            result
+          else
+            break nil
           end
         end
```

Comment:
Apart from backwards compatibility, is there any particular reason for keyword arguments? As I see it the keywords doesn't really add anything (and are completely ignored in the original implementation), and pondered adding it back ind, but deprecated.

Reply:
There, new version, how about that?

Reply:
> Apart from backwards compatibility, is there any particular reason for keyword arguments?

It's definitely not obvious anywhere in Armature (I didn't really expect anyone but me to be using it), but the keyword arguments are useful for tracing. I keep forgetting to extract to my OpenTelemetry shard, but the code I end up copy/pasting to every app using OpenTelemetry looks like this (based on [this method](https://github.com/jgaskins/opentelemetry/blob/abdb9632f1f974166ace458393e87d9ba68ed359/src/integrations/armature.cr#L27-L38)):

```crystal
module Route
  def on(**kwargs)
    previous_def do |value|
      capture = kwargs.first
      if span = top_otel_http_span
        match_path = (span["http.route"] ||= "").as(String)
        match_path += "/:#{capture}"
        span["http.route"] = match_path
        span["match.value.#{capture.to_s}"] = value
      end

      yield value
    end
  end
end
```

Then if my routes look like this:

```crystal
r.on "categories" do
  r.on category_id: UUID do |category_id|
    r.on "items" do
      r.on item_id: UUID do |item_id|
        r.get do
          # ...
        end
      end
    end
  end
end
```

… then my OTel spans have attributes like this:

| Attribute | Value |
|-|-|
| `http.route` | `"/categories/:category_id/items/:item_id.get"` |
| `match.value.category_id` | `"d81bcb1e-b2fa-40a3-aaf4-d1e8f19868f1"` |
| `match.value.item_id` | `"1cbd975d-e832-471b-8f33-c3a28a698e64"` |

That way, if I find a performance issue where these path segments are a factor, it'll show up in [Honeycomb's "Bubble Up" feature](https://www.honeycomb.io/bubbleup).

Reply:
Oh. You'd need to update it to support multiple arguments, but now you only have to override one `#on`.

I tried making `#is` use `#on` to avoid duplication, but `#root` got a bit in the way. That would've meant one less method to override.

I'll remove the deprecated annotation. And today I'll actually commit it before pushing :facepalm:.